hudi-spark-datasource/hudi-spark/src/test/scala/org/apache/hudi/functional/TestAutoKeyGeneration.scala

```diff
@@ -0,0 +1,180 @@
+    val compactionEnabled = if (tableType == HoodieTableType.MERGE_ON_READ) "true" else "false"
+    val writeOptions = commonOpts ++ Map(
+      DataSourceWriteOptions.TABLE_TYPE.key -> tableType.name(),
+      HoodieCompactionConfig.INLINE_COMPACT.key -> compactionEnabled
+    )
+    // Bulk insert first.
+    val records = recordsToStrings(dataGen.generateInserts("001", 5))
+    val inputDF: Dataset[Row] = spark.read.json(jsc.parallelize(records, 2))
+    inputDF.write.format("hudi").partitionBy("partition")
+      .options(writeOptions)
+      .option(DataSourceWriteOptions.OPERATION.key, DataSourceWriteOptions.BULK_INSERT_OPERATION_OPT_VAL)
+      .mode(SaveMode.Overwrite).save(basePath)
+    metaClient = HoodieTableMetaClient.builder.setConf(storageConf).setBasePath(basePath).build
+    assertEquals(9, metaClient.getTableConfig.getTableVersion.versionCode())
+    // Ensure no key fields are set.
+    assertTrue(metaClient.getTableConfig.getRecordKeyFields.isEmpty)
+    // 10 inserts.
+    for (i <- 0 until 10) {
+      val records = recordsToStrings(dataGen.generateInserts("$i%03d", 5))
+      val inputDF: Dataset[Row] = spark.read.json(jsc.parallelize(records, 2))
+      inputDF.write.format("hudi").options(writeOptions)
+        .option(DataSourceWriteOptions.OPERATION.key, DataSourceWriteOptions.INSERT_OPERATION_OPT_VAL)
+        .mode(SaveMode.Append).save(basePath)
+    }
+    // 10 updates.
+    for (i <- 10 until 20) {
+      val inputDF = spark.read.format("hudi").load(basePath)
+        .withColumn("weight", col("weight") * 2)
+      inputDF.write.format("hudi").partitionBy("partition")
+        .options(writeOptions)
+        .option(DataSourceWriteOptions.OPERATION.key, DataSourceWriteOptions.UPSERT_OPERATION_OPT_VAL)
+        .mode(SaveMode.Append).save(basePath)
+    }
+    // 10 deletes.
+    for (i <- 20 until 30) {
+      val inputDF = spark.read.format("hudi").load(basePath).limit(1)
+      inputDF.write.format("hudi").partitionBy("partition")
+        .options(writeOptions)
+        .option(DataSourceWriteOptions.OPERATION.key, DataSourceWriteOptions.DELETE_OPERATION_OPT_VAL)
+        .mode(SaveMode.Append).save(basePath)
+    }
+    // Validate configs.
+    metaClient = HoodieTableMetaClient.reload(metaClient)
+    assertTrue(metaClient.getTableConfig.getRecordKeyFields.isEmpty)
+    // Validate all records are unique.
+    val numRecords = spark.read.format("hudi").load(basePath).count()
+    assertEquals(45, numRecords)
+    // Validate clean, archive, and clustering operation exists.
+    assertFalse(metaClient.getActiveTimeline.getCleanerTimeline.getInstants.isEmpty)
+    assertFalse(metaClient.getArchivedTimeline.getInstants.isEmpty)
+    assertFalse(metaClient.getActiveTimeline.getCompletedReplaceTimeline.getInstants.isEmpty)
+    if (tableType == HoodieTableType.MERGE_ON_READ) {
+      assertFalse(metaClient.getActiveTimeline.getCommitsAndCompactionTimeline.empty())
```

Comment:
we should be able to ascertain exact number of compaction commits. 
same for clean commits and replace commits above as well instead of just checking for `~isEmpty`